include/deal.II/physics/notation.h

```diff
@@ -0,0 +1,619 @@
+                      const FullMatrix<Number> &mtrx);
+
+
+      /**
+       * Convert a compressed matrix to its equivalent rank-2 tensor.
+       */
+      template<typename Number>
+      static void
+      roll_to_tensor (Tensor<2,dim,Number> &t,
+                      const FullMatrix<Number> &mtrx);
+
+
+      /**
+       * Convert a compressed matrix to its equivalent rank-2 symmetric tensor.
+       */
+      template<typename Number>
+      static void
+      roll_to_tensor (SymmetricTensor<2,dim,Number> &st,
+                      const FullMatrix<Number> &mtrx);
+
+
+      /**
+       * Convert a compressed matrix to its equivalent rank-3 tensor.
+       *
+       * @note Based on the size of the matrix @p mtrx, some of the
+       * components of @p t may be interpreted as having symmetric
+       * counterparts. This is the reverse of the operation explained
+       * in the documentation of the counterpart unroll_to_matrix()
+       * function.
+       */
+      template<typename Number>
+      static void
+      roll_to_tensor (Tensor<3,dim,Number> &t,
+                      const FullMatrix<Number> &mtrx);
+
+
+      /**
+       * Convert a compressed matrix to its equivalent rank-4 tensor.
+       */
+      template<typename Number>
+      static void
+      roll_to_tensor (Tensor<4,dim,Number> &t,
+                      const FullMatrix<Number> &mtrx);
+
+
+      /**
+       * Convert a compressed matrix to its equivalent rank-4 symmetric tensor.
+       */
+      template<typename Number>
+      static void
+      roll_to_tensor (SymmetricTensor<4,dim,Number> &st,
+                      const FullMatrix<Number> &mtrx);
+
+
+      /**
+       * A generic helper function that will convert a compressed vector
+       * to its equivalent @p TensorType.
+       */
+      template<typename TensorType, typename Number>
+      static TensorType
+      roll_to_tensor (const Vector<Number> &vec);
+
+
+      /**
+       * A generic helper function that will convert a compressed matrix
+       * to its equivalent @p TensorType.
+       */
+      template<typename TensorType, typename Number>
+      static TensorType
+      roll_to_tensor (const FullMatrix<Number> &vec);
+//@}
+
+    private:
+
+      /**
+       * @name Indexing
+       */
+//@{
+
+      /**
+       * Return the tensor indices <code><row, column></code>
+       * associated with a condensed component index. The
+       * @p symmetric flag indicates whether or not the
+       * @p component_n index is associated with a tensor that
+       * has symmetric entries.
+       */
+      static std::pair<unsigned int, unsigned int>
+      indices_from_component (const unsigned int component_n,
+                              const bool         symmetric);
+
+      /**
+       * Return the scaling factor to be applied to the
+       * entry in the condensed vector.
+       */
+      static double
+      vector_component_factor (const unsigned int component_i,
```

Comment:
does it need to be public interface? same below for `matrix_component_factor`

Reply:
No, these are internal functions used to construct the matrices and vectors.

Reply:
sure, i completely missed `private:` thing.